### DISCUSSION TITLE
Opened Scenes as dockable panel?

### DISCUSSION THREAD

**Discussion by seppoday** [godotengine/godot-proposals#6499] [Editor]
Lately I was kinda fascinated by Arc browser and I've got idea. What if we would move open scene tabs to its own dockable panel?

There are few pros and cons with that approach:

**Pros:**
1. You can move dockable panel
2. You can resize dockable panel
3. It reads better (questionable...) :P
4. It kinda works better at full height and with longer scene names? You can open a lot of scenes. Which might be not great thing to do. But some of us do it anyway... Currently you can open 6-8 scenes and bar will get full and those arrows will appear. With this approach you can expand it vertically or just scroll through list or filter scenes.
5. This approach is expandable. For example we might add filter by node type or add grouping for organized people (like you can create folders etc... for example) and use "close group" or have 2 scenes in split-view in the future or store all scenes from project in that panel and you can open them from panel (no need to find them in filesystem) etc... Just some ideas.
6. We could add "pin scene" feature
7. We could add search field
8. We could also add contextual menu (3 dots) with for example "close all scenes" or "close unpinned" option
9. It creates more room for main editor window
10. It also creates more room in Script Editor and Asset Library window because there is no tabs above it. Also looks kinda cleaner because of that and there might be less confusion in script editor when you currently can have one scene opened on top and edit script from other scene.
11. I feel like it creates more flexibility in the future for improving editor UI?
12. You can make floating panel and move it to another monitor (double monitor support)
13. It kinda works nice with 4k monitors when you go with 2 panels on right or left side (full height)
14. If you just work on one scene for a moment. You can just close panel and save space.
15. (probably) it might work well with addon creators? They might have access to that panel (or not?)
16. It might enable some things in the future. Like easier drag and drop opened scenes into other panels for some reason?
18. Probably few more...

**Cons:**
1. People are kinda familiar with tabs, so it might (or not) be confusing at first?
2. Panel eats up space. So we kinda get more space above main editor window, but we need to have panel opened at side
3. With panel closed you might not know in which scene you are (?). Maybe adding some grey-ish text with current opened scene might help if that's a big problem?
4. People might confuse that panel with "All scenes in project" and not "Opened scenes"? Maybe it should be named "Opened scenes" also it probably should be opened by default when you open project.
5. There are probably few more cons with how it would work UX wise, but I am not PRO with Godot Engine. So please comment your thoughts.

What do you think?

Top right corner.
1. New project created with empty scene
![image](https://user-images.githubusercontent.com/62170071/225049284-05b65863-cfdb-4bd9-9766-270811c356bb.png)

2. Only normal scenes
![image](https://user-images.githubusercontent.com/62170071/225051366-eeacab64-64a1-4ad2-a16f-c2022b685c75.png)

3. Some pinned and some normal scenes
![image](https://user-images.githubusercontent.com/62170071/225051480-59295c81-e103-43f6-aae0-1ffafbda9ee4.png)

4. Only pinned
![brave_xpmdE4aps1](https://user-images.githubusercontent.com/62170071/225045065-6ef0016b-077f-4ac2-bc55-bd64b56f5726.png)

5. Full height
![image](https://user-images.githubusercontent.com/62170071/225046905-209835d6-e78b-4ec9-abca-b8ec9221fc85.png)

6. Full height and double panel on one side
![image](https://user-images.githubusercontent.com/62170071/225050087-f4bd10e7-ad9c-48bc-835e-a92617f3035f.png)

7. Script editor view
![obraz](https://user-images.githubusercontent.com/62170071/225251385-56b6f08d-95ac-43fb-9c5f-612b80947b98.png)

PS: Close symbol might only show when in hover state. Dunno...

**Comment by poyribn** [comment]
i think that having the list of opened scenes disconnected from the scene viewport will make godot harder to use, because beginners will miss it and think that they can only open one scene at a time: when they want to open another scene they recently edited, they will hunt it in the filesystem panel.

having the tabs, but with a 'toggle vertical tabs' button (like in microsoft edge) is almost as powerful but much easier to understand — a trade-off that is better than your original solution and better than godot's current horizontal-only tabs

**Reply by zinnschlag** [reply]
The current setup also confuses beginners. When in script mode the editor still shows the scene tab even if the edited script is completely unrelated to it. Makes sense for an experienced user. But it is very unintuitive at first. There have been plenty of proposals and confused comments related to this issue. I think removing the tab bar altogether will do more good than harm.

**Reply by seppoday** [reply]
I think both concerns are valid. Both implementations have some pros and cons. Question is which one of them might be better for the community overall.

**Reply by poyribn** [reply]
i think that we should still have the tabs, but remove the possibility of having script tabs in the scene editor:

the scene editor should have its own set of tabs that can only have scenes,
and the script editor should have its own set of tabs that can only have scripts.

**Reply by zinnschlag** [reply]
This would make it impossible to change scenes while in the script editor. Major loss of QoL.

**Comment by zinnschlag** [comment]
Shouldn't this be a proposal instead of a discussion? It has enough details to be actionable.

**Reply by seppoday** [reply]
Unfortunately I lack knowledge to implement that feature. That's why I've opened discussion. If it might help anyone I am happy to open proposal in issue tab.

> "If you have a more general idea for a feature but are not well versed in Godot's architecture or do not possess the necessary knowledge to implement it in the engine, feel free to open a [discussion](https://github.com/godotengine/godot-proposals/discussions/new) instead of an [issue](https://github.com/godotengine/godot-proposals/issues/new/choose)."

**Reply by seppoday** [reply]
Also...it is kinda huge change to basically all users. So I think it should be discussed thoroughly first. I don't want to push implementing something because only I think it might be good for community.

**Comment by eon-s** [comment]
I think it will be confusing, since scenes are not necessarily related with each other.

But with a system like this https://github.com/godotengine/godot-proposals/discussions/4091 could make more sense in context.

Like, show a list of instanced scenes so you can quick edit, hide or highlight those in the current viewport.

**Reply by seppoday** [reply]
One could argue that in tab mode they are also not related to each other. Tabs in your browser are also not related to each other. They are only webpages. In here we only have scenes. They do not need to be related.

Idea from #4091 looks nice and those two might fit together. Interesting idea.

**Comment by azur-wolve** [comment]
by default do it with tabs
but allow an option to turn into the vertical list you come up with

point 15 makes me think that having a vertical panel with a list allows to use separated sections
this could be great to have an upper section above the normal opened scenes, "editor scenes" that really are custom workspaces or editors.
maybe could be of some good use, but still already can have custom workspaces and editors for addons, so idk

**Comment by poyribn** [comment]
after some thinking, switching from the scene tabs to a 'RecentScenes' dock will make Godot's design more consistent, and easier to use because the panel is hidden from beginners who don't need it yet.

the RecentScenes panel layout would be very similar to the History panel's, and the RecentScenes tab could be next to the History tab by default.

**Comment by djrain** [comment]
Whatever the replacement for scene tabs is (and the current scene tabs are really terrible btw), I hope that it's fully customizable / replaceable via plugin. Clearly, people have different preferences on this, maybe even per-project, and there are a lot of options that could work well.
